cpp/ql/src/Diagnostics/Files.ql

```diff
@@ -0,0 +1,10 @@
+/**
+ * @name Total source files
+ * @description The total number of source files.
+ * @kind metric
```

Comment:
Is `@kind metric` all the metadata we need for Code Scanning to present the results of these queries in the right place? I note that `@kind metric` is already in use by LGTM, and [the LGTM documentation](https://help.semmle.com/lgtm-enterprise/user/help/writing-custom-queries.html) specifies that such queries have two additional mandatory properties. We have existing `@kind metric` queries for Python and JavaScript in this repo.

Reply:
Oh well spotted. I reused the existing `@kind metric` deliberately, but will look into making those properties optional in the toolchain. If we're not adding these queries to the LGTM suites then we don't need to make LGTM handle them.

Reply:
> Is `@kind metric` all the metadata we need for Code Scanning to present the results of these queries in the right place?

It sounds like we'll need a tag and/or a query suite to identify the metrics to use in Code Scanning. Any decision on what to use here, or should we just add it later when the rest of the toolchain is further along?

Reply:
I can see the drawbacks of reusing `@kind metric`: we have to change the rules for additional properties and possibly change how the suites are produced. What are the advantages?

Reply:
- Go with just `@kind metric` for now, I think that's enough for you to proceed :)
- The contents of the Code Scanning suites are determined by https://github.com/github/codeql/blob/main/misc/suite-helpers/code-scanning-selectors.yml. Regardless of what name we use here, we'll have to update that config to ship these new types of queries to Code Scanning. The suite format gives us the ability to include by `kind`.
- The LGTM suite generation script only includes queries that also have `metricType`.
- The existing meta-metric queries for JS and Python can easily be excluded by filtering out `@tags meta`, or just editing their metadata if those teams are happy with it (LGTM and Code Scanning don't include them anyway).
- I think metric is the simplest name for this property, and it gives scope to reuse the metric/treemap queries we already have that fit this purpose. That said, the choice shouldn't block you here: if we encounter trouble making the toolchain work, or need to do a final find-replace with a new name, that's on my team to fix.

Reply:
> The existing meta-metric queries for JS and Python can easily be excluded by filtering out `@tags meta`

I'd be uncomfortable with a query suite defined as exclusion-filter, i.e. "everything except queries with tags x, y, z". A simple typo in a tag could cause a query to be picked up by accident, making its way into production.

Would you please consider an inclusion-filter instead, for example,
```yaml
- include:
    kind:
    - metric
    tags contain:
    - diagnostic
```

(sorry for hijacking the thread, it seemed related to this)

Reply:
I don't know the technicalities of how big of an effort it would be to invent a new _kind_ vs. the effort of changing the docs. I also don't know our overall philosophy of how much we try to reuse existing _kinds_ vs. create new ones.

If we do reuse the _kind_, I'm very much in support of Asger's proposal to let the suites be determined by inclusion filters rather than exclusion filters. It seems like too much of a footgun to require every future LGTM-style `metric` query to add `@tags meta`. Authors of such queries will be testing the LGTM-style functionality, and they won't notice that Code Scanning suddenly starts including extra diagnostic data in its logs because omitted a tag.

Reply:
Since we're already going with "summary" for the path name and `@id`, should we use `summary` as the tag, rather than `diagnostic`?

Reply:
Inclusion filter sounds sensible. I think `summary` is a good tag name for now since we eventually plan to have `diagnostic` as a separate kind (for extractor diagnostics), so let's go with that. If you want both of those tag names supported later it's easy to add.